nixos/modules/services/mail/stalwart-mail.nix

```diff
@@ -0,0 +1,104 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.stalwart-mail;
+  configFormat = pkgs.formats.toml { };
+  configFile = configFormat.generate "stalwart-mail.toml" cfg.settings;
+  dataDir = "/var/lib/stalwart-mail";
+
+in {
+  options.services.stalwart-mail = {
+    enable = mkEnableOption (mdDoc "the Stalwart all-in-one email server");
+    package = mkPackageOptionMD pkgs "stalwart-mail" { };
+
+    settings = mkOption {
+      inherit (configFormat) type;
+      default = { };
+      description = mdDoc ''
+        Configuration options for the Stalwart email server.
+        See <https://stalw.art/docs/> for available options.
+
+        By default, the module is configured to store everything locally.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    # Default config: all local
+    services.stalwart-mail.settings = {
+      global.tracing.method = mkDefault "stdout";
+      queue.path = mkDefault "${dataDir}/queue";
+      report.path = mkDefault "${dataDir}/reports";
+      store.db.path = mkDefault "${dataDir}/data/index.sqlite3";
+      store.blob.type = mkDefault "local";
+      store.blob.local.path = mkDefault "${dataDir}/data/blobs";
+      resolver.type = mkDefault "system";
+    };
+
+    systemd.services.stalwart-mail = {
+      wantedBy = [ "multi-user.target" ];
+      after = [ "local-fs.target" "network.target" ];
+
+      preStart = ''
+        mkdir -p ${dataDir}/{queue,reports,data/blobs}
+      '';
+
+      serviceConfig = {
+        ExecStart =
+          "${cfg.package}/bin/stalwart-mail --config=${configFile}";
+
+        # Base from template resources/systemd/stalwart-mail.service
+        Type = "simple";
+        LimitNOFILE = 65536;
+        KillMode = "process";
+        KillSignal = "SIGINT";
+        Restart = "on-failure";
+        RestartSec = 5;
+        StandardOutput = "syslog";
+        StandardError = "syslog";
+        SyslogIdentifier = "stalwart-mail";
+
```

Comment:
Should we register some static user and group ID in `misc/ids.nix` then?

It might also make sense with the Maildir storage mode, storing a lot of files
which are not instant to chown by systemd when the service starts.

However the DynamicUser approach does make more sense in the setup using S3
and an external database.

Reply:
(Also, I do not see the necessity of `settings.server.run-as.user = "stalwart-mail"` in either case)?

Reply:
I think we can leave it dynamic with a user name, it gives us the best of both world I think.
you can use tmp rules or sudo su user, but you still get the benefit of dynamic user.

for the run-as.user, I'm not sure about that one, it says in the docs that the service runs as the user, I thought it might as well be the correct user. But maybe it doesn't matter either way.

Reply:
Added the `User` key while keeping `DynamicUser` and amended.

The `run-as.user` seems ot be mainly for starting as root to listen on the
privileged ports then drop those privileges:
https://stalw.art/docs/configuration/general/#run-as-user

Though it's not necessary here since we're doing that through
`CAP_NET_BIND_SERVICE`, and Systemd already starts the service as the correct
user.